File: docs/JAMIA/comparisons/RPython/TM/30132282_str_on_datetime_series.py
# ID: SO#30132282
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: .str accessor on datetime Series. Wrong accessor for column type.
# Reproduction status: Live
# Type system catch: `.str` accessor rejects `Temporal` column; requires `string`

import pandas as pd

dates = pd.to_datetime(pd.Series(['20010101', '20010331']), format='%Y%m%d')
print(f"dtype: {dates.dtype}")

# BUG: .str accessor on datetime Series
try:
    result = dates.str.slice(0, 10)
    print(result)
except AttributeError as e:
    print(f"Crash: {e}")

File: docs/JAMIA/comparisons/RPython/TM/30132282_str_on_datetime_series.ts
/**
 * RPython SO#30132282 — datetime to string with series in pandas
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In pandas, .str accessor on a datetime Series crashes. The user wants to
 * format the dates as strings but uses the wrong accessor for the dtype.
 *
 * In tidy-ts, dates stored as strings are typed as string — string methods
 * work. Dates stored as numbers can't have string methods called on them.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { date_str: "2001-01-01", amount: 100 },
  { date_str: "2001-03-31", amount: 200 },
]);

// date_str is string — string operations work in mutate
const formatted = df.mutate({
  year: (r) => r.date_str.slice(0, 4),
});

// Numeric operations on string column are caught
// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("date_str"));

formatted.print();
